File: milestone 2/python_modules/module03/ex3/ft_achievement_tracker.cpp
#include <iostream>
#include <string>
#include <vector>
#include <cstdlib>
#include <ctime>
using namespace std;

class Player {
public:
    string name;
    vector<string> achievements;

    Player(string n, vector<string> a) : name(n), achievements(a) {}
};

vector<string> get_random_achievements(vector<string>& all_achievements)
{
    vector<string> result;
    vector<string> pool = all_achievements;

    int max = all_achievements.size();

    int count = rand() % max;

    for (int i = 0; i < count; i++)
    {
        int index = rand() % pool.size();
        result.push_back(pool[index]);
        pool.erase(pool.begin() + index);
    }

    if (result.empty())
        return {"set()"};

    return result;
}

int main(void)
{
    srand(time(0));

    cout << "=== Achievement Tracker System ===" << endl;

    vector<string> achievements = {
        "Crafting Genius", "Strategist", "World Savior", "Speed Runner", "Survivor",
        "Master Explorer", "Treasure Hunter", "Unstoppable", "First Steps",
        "Collector Supreme", "Untouchable", "Sharp Mind", "Boss Slayer", "Hidden Path Finder"
    };

    vector<Player> players = {
        Player("Alice",   get_random_achievements(achievements)),
        Player("Bob",     get_random_achievements(achievements)),
        Player("Charlie", get_random_achievements(achievements)),
        Player("Dylan",   get_random_achievements(achievements))
    };

    for (Player p : players)
    {
        cout << "\nPlayer " << p.name << ": ";
        for (int i = 0; i < p.achievements.size(); i++)
        {
            cout << "{'" << p.achievements[i] << "'}";
            if (i < p.achievements.size() - 1)
                cout << ", ";
            else if (i == p.achievements.size() - 1)
                cout << endl;
        }
    }

    return 0;
}

File: milestone 2/python_modules/module03/ex3/ft_achievement_tracker.py
import random

def gen_player_achievements():
    achievement_list: list = [
        'Crafting Genius', 'Strategist', 'World Savior', 'Speed Runner', 'Survivor',
        'Master Explorer', 'Treasure Hunter', 'Unstoppable', 'First Steps',
        'Collector Supreme', 'Untouchable', 'Sharp Mind', 'Boss Slayer', 'Hidden Path Finder'
    ]
    count: int = random.randint(1, len(achievement_list))
    return set(random.sample(achievement_list, k=count))


if __name__ == "__main__":
    print("=== Achievement Tracker System ===\n")

    players: dict = {
        "Alice": gen_player_achievements(),
        "Bob": gen_player_achievements(),
        "Charlie": gen_player_achievements(),
        "Dylan": gen_player_achievements(),
    }

    for name, achievements in players.items():
        print(f"Player {name}: {achievements}")

    all_achievements: set = set().union(*players.values())
    print(f"\nAll distinct achievements: {all_achievements}")

    common: set = set.intersection(*players.values())
    print(f"Common achievements: {common}")

    print()
    for name, achievements in players.items():
        others: set = set.union(*[ach for n, ach in players.items() if n != name])
        print(f"Only {name} has: {achievements.difference(others)}")

    print()
    for name, achievements in players.items():
        print(f"{name} is missing: {all_achievements.difference(achievements)}")